test/extended/machines/cluster.go

```diff
@@ -100,4 +119,293 @@ var _ = g.Describe("[sig-node] Managed cluster", func() {
+		o.Expect(err).NotTo(o.HaveOccurred())
+		for i := range mcList.Items {
+			mcName := mcList.Items[i].GetName()
+			for prefix := range expectedReboots {
+				if strings.HasPrefix(mcName, fmt.Sprintf("rendered-%s-", prefix)) {
+					expectedReboots[prefix] += 1
+				}
+			}
+		}
+
+		// List nodes, set actual and expected number of reboots
+		for _, node := range nodes.Items {
+			// Examine only the nodes which are available at the start of the test
+			if !slices.Contains(staticNodeNames, node.Name) {
+				continue
+			}
+			expectedRebootsForNodeRole, ok := expectedReboots[getNodeRole(&node)]
+			o.Expect(ok).To(o.BeTrue())
+			expected[node.Name] = expectedRebootsForNodeRole
+			nodeReboots, err := getNumberOfBootsForNode(oc.KubeClient(), node.Name, 0)
+			if err != nil {
+				errs = append(errs, err)
+			}
+			actual[node.Name] = nodeReboots
+		}
+
+		o.Expect(errs).To(o.BeEmpty())
+		o.Expect(actual).To(o.Equal(expected))
+	})
 })
+
+// getNodeRole reads node labels and returns either "worker" or "master"
+func getNodeRole(node *corev1.Node) string {
+	if _, ok := node.Labels["node-role.kubernetes.io/worker"]; ok {
+		return "worker"
+	}
+	if _, ok := node.Labels["node-role.kubernetes.io/master"]; ok {
+		return "master"
+	}
+	return "worker"
+}
+
+func getNumberOfBootsForNode(kubeClient kubernetes.Interface, nodeName string, attempt int) (int, error) {
+
+	// Run journalctl to collect a list of boots
+	command := "exec chroot /host journalctl --list-boots"
+	isTrue := true
+	zero := int64(0)
+	ctx := context.Background()
+	name := fmt.Sprintf("list-boots-%s-%d", nodeName, attempt)
+	pod, err := kubeClient.CoreV1().Pods(namespace).Create(context.Background(), &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: corev1.PodSpec{
+			Tolerations: []corev1.Toleration{
+				{
+					Effect:   "NoSchedule",
+					Key:      "node-role.kubernetes.io/master",
+					Operator: corev1.TolerationOpExists,
+				},
+			},
+			HostPID:            true,
+			RestartPolicy:      corev1.RestartPolicyNever,
+			NodeName:           nodeName,
+			ServiceAccountName: serviceAccountName,
+			Volumes: []corev1.Volume{
+				{
+					Name: "host",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/",
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name: "list-boots",
+					SecurityContext: &corev1.SecurityContext{
+						RunAsUser:  &zero,
+						Privileged: &isTrue,
+					},
+					Image: image.ShellImage(),
+					Command: []string{
+						"/bin/bash",
+						"-c",
+						command,
+					},
+					TerminationMessagePolicy: corev1.TerminationMessageFallbackToLogsOnError,
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							MountPath: "/host",
+							Name:      "host",
+						},
+					},
+				},
+			},
+		},
+	}, metav1.CreateOptions{})
+	if apierrors.IsAlreadyExists(err) {
+		return getNumberOfBootsForNode(kubeClient, nodeName, attempt+1)
```

Comment:
Is there a limit on recursive attempts?

Reply:
Good point, limited attempts to 10, checked at the top of the procedure